blockchain/vm/contracts.go

```diff
@@ -825,3 +862,501 @@ func (c *validateSender) validateSender(input []byte, picker types.AccountKeyPic
+	// > Output is an encoding of addition operation result - single G2 point (`256` bytes).
+	if len(input) != 512 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	var err error
+	var p0, p1 *bls12381.G2Affine
+
+	// Decode G2 point p_0
+	if p0, err = decodePointG2(input[:256]); err != nil {
+		return nil, err
+	}
+	// Decode G2 point p_1
+	if p1, err = decodePointG2(input[256:]); err != nil {
+		return nil, err
+	}
+
+	// No need to check the subgroup here, as specified by EIP-2537
+
+	// Compute r = p_0 + p_1
+	r := new(bls12381.G2Affine)
+	r.Add(p0, p1)
+
+	// Encode the G2 point into 256 bytes
+	return encodePointG2(r), nil
+}
+
+// bls12381G2Mul implements EIP-2537 G2Mul precompile.
+type bls12381G2Mul struct{}
+
+// GetRequiredGasAndComputationCost returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G2Mul) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+	return params.Bls12381G2MulGas, params.Bls12381G2MulComputationCost
+}
+
+func (c *bls12381G2Mul) Run(input []byte, contract *Contract, evm *EVM) ([]byte, error) {
+	// Implements EIP-2537 G2MUL precompile logic.
+	// > G2 multiplication call expects `288` bytes as an input that is interpreted as byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes).
+	// > Output is an encoding of multiplication operation result - single G2 point (`256` bytes).
+	if len(input) != 288 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	var err error
+	var p0 *bls12381.G2Affine
+
+	// Decode G2 point
+	if p0, err = decodePointG2(input[:256]); err != nil {
+		return nil, err
+	}
+	// 'point is on curve' check already done,
+	// Here we need to apply subgroup checks.
+	if !p0.IsInSubGroup() {
+		return nil, errBLS12381G2PointSubgroup
+	}
+	// Decode scalar value
+	e := new(big.Int).SetBytes(input[256:])
+
+	// Compute r = e * p_0
+	r := new(bls12381.G2Affine)
+	r.ScalarMultiplication(p0, e)
+
+	// Encode the G2 point into 256 bytes
+	return encodePointG2(r), nil
+}
+
+// bls12381G2MultiExp implements EIP-2537 G2MultiExp precompile.
+type bls12381G2MultiExp struct{}
+
+// GetRequiredGasAndComputationCost returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G2MultiExp) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+	// Calculate G2 point, scalar value pair length
+	k := len(input) / 288
+	if k == 0 {
+		// Return 0 gas for small input length
+		return 0, 0
+	}
+	// Lookup discount value for G2 point, scalar value pair length
+	var discount uint64
+	if dLen := len(params.Bls12381MultiExpDiscountTable); k < dLen {
+		discount = params.Bls12381MultiExpDiscountTable[k-1]
+	} else {
+		discount = params.Bls12381MultiExpDiscountTable[dLen-1]
+	}
+	// Calculate gas and return the result
+	return (uint64(k) * params.Bls12381G2MulGas * discount) / 1000,
+		710 * (uint64(k) * params.Bls12381G2MulGas * discount) / 1000 / 16
```

Comment:
Same here, let's go for `uint64(k) * params.Bls12381G2ComputationCost * discount) / 1000`

Reply:
Fixed https://github.com/kaiachain/kaia/pull/111/commits/66e1d18b02dd1004680a42fe1cb4a5360fa0f567